### ISSUE TITLE
Proposal: simplifying options and shortcuts syntax

### ISSUE THREAD

**Issue by McJones** [YarnSpinnerTool/YarnSpinner#264]
# Rationale

Yarn Spinner began life as the tool needed for Night in the Woods and was inspired by Twine, as such its features and syntax grew as NitW grew from something akin to twine to its own thing.
We've now got a fair amount of syntax features in the language that replicate or complicate existing functionality already in the language.
As we move to YS 2.0 we can make breaking changes so now is the time to think about them.
This proposal is one of a few I'll be making around looking at potential options to trim out some of the feature and syntax duplication (as I see it) in Yarn Spinner (each will have its own proposal to avoid clashing on discussing), beginning with:

## Jumps and shortcuts

As it currently stands we have two different language constructs for presenting options to a user:

```
[[Some Text|NodeName]]

-> Some Text
    inner-text // or the equivalent text of what is inside NodeName node
```

From a users perspective these are identical, they are shown one or more potential selections to make, they make them, and the game continues as normal.
Even internally they function in a very similar manner.
From a non-extensive look at how people have been writing Yarn dialogue it looks like to me that the `->` shortcut syntax is by far the preferred means of offering choice in preference to making new nodes and using jumps.
Nodes seem to be used only for large story chunks.

This is I think a side-effect of the twine heritage.
Twine's syntax encourages lots of little story nodes which you chain together to present options and do branching, Yarn encourages large nodes with lots of little shortcuts inside and then jumps to another large node as the way of writing your story.
I would argue we are no longer twine-like in how people write dialogue in Yarn vs how they write it in Twine.

Additionally the `[` character is already in use for the markup and format functions which exist only in the context of a line.
The general way things are controlled in Yarn is through the `<<` and `>>` terminals, the `[[` only exists for jumps.

In short I think we get rid of one of these and roll its functionality into the other.

# The Proposal

This proposal suggests that we get rid of the `[[text|node]]` syntax and instead put its functionality as an optional component of the `->` syntax.
This would work by adding a new command to the end of the `->` statement.

This means:

```
[[Some Text|NodeName]]
```

Would become:

```
-> Some Text <<jump NodeName>>
```

The shortcut syntax already has support for expectating `<< >>` at the end, just that it currently is only for conditional expressions instead of commands.

The command itself would be a dedicated built-in one like `stop` that the system itself understands and handles without talking to the command handlers.
This does mean that any current uses of custom `jump` commands would be broken and would need to be flagged.
Some potential other names could be:

- jumpTo
- goto
- move
- loadNode

## Advantages

The main advantage here is we simplify how we talk about moving through dialogue, there is now only one way to do so.
This also frees up the `[]` symbols such that they are only used as part of the markup and not also flow control.

This also simplifies potential expansions to jumps such as the return statement spoken about in issue #92, we'd use a new keyword instead of `jump` in the command, perhaps `return`.

You can also now just jump using the command directly should you wish to force a move.
As `jump` would just be a special command you could sprinkle these in your story and have them move automatically.

## Disadvantages

This is a syntax breaking change.
As such any file written for Yarn Spinner 1.x.y would need to be rewritten for 2.x.y.
Luckily this is an easy change to detect and automatically process using the following steps:

1. Identify any `[[]]` statements
2. Work out if they are a `[[text|node name]]` or a `[[node name]]`
3. Replace the first category with `-> text <<jump node name>>`
4. Replace the second category with `<<jump node name>>`
5. Detect any commands that start with `jump`
6. Fire off warning for these commands as now conflicting

### Do nothing

The current solution isn't bad as such, just less than ideal, not making this change isn't exactly gonna hurt anything.

### Use `[[|]]` as the default

This would mean replacing the `->` syntax with the `[[]]` syntax.
While it could be done it personally feels worse to me but there is no actual advantage or disadvantage here over picking `->` just that the arrow syntax seems more popular.

### Invent an entirely new approach

Much like with the above there is nothing *wrong* with this, just more that it feels wrong to do so, why make a total change like that?

**Comment by desplesda** [comment]
A primary reason for the `[[ ]]` syntax was syntactic similarity to Twine. I don't feel like maintaining that similarity is a key goal for us at this point, so I've got no objection to a change like this. I may have one or two pieces of feedback as I think about this proposal, but it's got my vote.

**Comment by desplesda** [comment]
(It's also occurred to me that if we go ahead with a jump-and-return feature (as discussed in https://github.com/YarnSpinnerTool/YarnSpinner/issues/92), the syntax for it could look something like `<<jump NodeName return>>`.

**Comment by desplesda** [comment]
Two minor points: 

1. I'd suggest doing the potential jump-and-return syntax with a parameter on the `jump` command, i.e.:

```
<<jump NodeName return>>
```

This spares us from having to reserve an additional keyword.

2. The current syntax allows for options to be accumulated in a list, and presents them when the end of a node is reached. 

For example, consider the following:

```
Hey!

[[Here's Option 1!|Option1Selected]]

<<if $var == true>>
    [[Here's Option 2!|Option2Selected]]
<<endif>>

Well, bye!
```

This will present the two lines ("Hey!", "Well, bye!"), and then present options; if `$var` is true, two options will be shown, otherwise only one.

I'm not entirely sure how many users actually make use of this, but I'm curious to know how we'd replicate a similar behaviour in this proposed syntax.

**Comment by McJones** [comment]
1. eh potato, tomato. I don't really have strong opinions on the phrasing of the return, more that this is a natural location for it to be handled.

2. Is anyone using that? Because until 2 minutes ago I literally did not know that was the case. Regardless is it even desirable feature?
Also does it show the Well, bye! before or after the options, or is that skipped?
This just looks like really weird yarn to me...

That said assuming this is an absolutely desperate feature I would probably make the following change to the grammar:

```ebnf
line_statement
    : 
        line_formatted_text // text, interspersed with expressions
        line_condition? // a line condition
        hashtag*  // any number of hashtags
        (TEXT_NEWLINE|TEXT_COMMANDHASHTAG_NEWLINE) // the end of the line
    ;
```

to:

```ebnf
line_statement
    : 
        line_formatted_text // text, interspersed with expressions
        line_condition? // a line condition
        command_statement?
        hashtag*  // any number of hashtags
        (TEXT_NEWLINE|TEXT_COMMANDHASHTAG_NEWLINE) // the end of the line
    ;
```

Note this is without checking if this works, just roughly reading the antlr and making some assumptions...
Much like where we currently accept any line as having a condition we can make it so it can also have a command at the end, and we'd just throw errors if these are encountered outside of shortcuts.

This means the yarn would change from:

```
Hey!

[[Here's Option 1!|Option1Selected]]

<<if $var == true>>
    [[Here's Option 2!|Option2Selected]]
<<endif>>

Well, bye!
```

to

```
Hey!

-> Here's Option 1 <<jump Option1Selected>>
-> Here's Option 2 <<$var == true>><<jump Option2Selected>>

Well, bye!
```

**Comment by desplesda** [comment]
The sample replacement you gave doesn't have precisely the same behaviour as the original, but the overall idea makes sense - detect the option, and move it to the end of the node using the new syntax. 

I’m actually fine with breaking this feature in its more complex forms (like when an option is accumulated inside an `if` statement), pending standard documentation and explanation for those who are using it of course. I think we can detect this feature’s use by flagging options that are inside an if statement’s clauses, and issuing a warning that the syntax it’s being replaced with will be presented immediately, or turning it into an error.

(It also occurs to me that this feature is currently buggy because of the interaction between accumulated options and shortcut options, which both use the _same_ pending options list.)

tl;dr: I propose detecting instances of this pattern, moving it where it's easy to do, and issuing an error where it's not.

**Comment by WeslomPo** [comment]
In my similar language inspired by YS, I rewrite syntax to
```
-> Some Text | NextNodeId // [[Some Text | NextNodeId]]

This syntax like previous, and there no additional command like `jump` (there are already reserved `|` symbol)

-> Some Text
    Text will be shown when you select this option
-| MultilineTransitionNextNodeId // [[-> ... [[MultilineTransitionNextNodeId]]

-| NoOptionTransitionNextNodeId // [[NoOptionTransitionNextNodeId]]

 I changed << ... >> to ~{ ... } and you can write conditions like this

-> Some Text | NextNodeId ~if (conditionHere)
-> Some Text | NextNodeId ~unless (conditionHere) // same as `not conditionHere`
-| NoOptionWithCondition ~if (...)

```

**Comment by desplesda** [comment]
What does the `-|` syntax do? Act like the current `[[NodeName]]` jumping syntax?

**Comment by WeslomPo** [comment]
Yes and no, without any actions jump syntax. There 2 situations:
```-> Option text
    Text message or something
-| Jump
```

and 
```
-| JumpWithoutAction
```

**Comment by desplesda** [comment]
Makes sense!

I think we'll probably keep the `<<jump>>` command, since doing that means we don't need to add support in the lexer for another symbol, and the `jump` keyword is a little more directly understandable. This is a very cool alternate approach, though!

**Comment by desplesda** [comment]
The relevant changes to the grammar have been added as of fa3290fac3f512ed482eba886dd231784098ad40. I'm now looking into adding support in the updater for some degree of automatic conversion from the previous syntax.

**Comment by desplesda** [comment]
The upgrader is now added, and re-writes old-style (`[[`) options into what were previously called "shortcut options" (`->`), as well as rewriting the previous syntax for jumps (`[[Destination]]`) into a `jump` command (`<<jump Destination>>`). The upgrader preserves all hashtags on an option, and also examines any `if` statements that an option might be embedded in and rewrites them into a line condition.

Before:

```
[[Option 1|DestOption1]] #line:a1b2c3
[[Option 2|DestOption2]]

Here's a line. It runs before options appear.

<<if 1 + 1 == 2>>
    [[Option 3|DestOption3]] #line:d4e5f6
<<elseif 1 + 1 == 3>>
    <<if 1 + 1 == 4>>
        [[Option 4|DestOption4]]
    <<endif>>
<<else>>
    [[Option 5|DestOption5]]
<<endif>>
```

After:

```
// Option "Option 1" moved to the end of this node
// Option "Option 2" moved to the end of this node

Here's a line. It runs before options appear.

<<if 1 + 1 == 2>>
    // Option "Option 3" moved to the end of this node
<<elseif 1 + 1 == 3>>
    <<if 1 + 1 == 4>>
        // Option "Option 4" moved to the end of this node
    <<endif>>
<<else>>
    // Option "Option 5" moved to the end of this node
<<endif>>

-> Option 1 #line:a1b2c3
    <<jump DestOption1>>
-> Option 2
    <<jump DestOption2>>
-> Option 3 <<if (1 + 1 == 2)>> #line:d4e5f6
    <<jump DestOption3>>
-> Option 4 <<if !(1 + 1 == 2) && (1 + 1 == 3) && (1 + 1 == 4)>>
    <<jump DestOption4>>
-> Option 5 <<if !(1 + 1 == 3) && !(1 + 1 == 2)>>
    <<jump DestOption5>>
```

With this added to `develop`, I'm marking this issue as done. Thanks to everyone who helped figure this proposal out!
